File: scripts/plot/python/rq1/analyze_ablation_jsonl.py
#!/usr/bin/env python
"""
Quick analyzer for ablation JSONL outputs.

It streams the JSONL file and reports:
  - numeric accuracy for baseline/ablated texts per (mode, step)
  - flip rates: baseline correct -> ablated wrong and baseline wrong -> ablated correct
  - teacher-forced log-prob deltas per (mode, step):
        * teacher_forced_delta_sum       (Δ log p(gold sequence) = base - ablt)
        * delta_logp_final_token         (Δ log p(final gold token) = base - ablt)

Assumes answers are of the form:
    {answer} <eos>
and gold_answer is a string containing the gold number (e.g., "70000").

Usage:
  python scripts/plot/python/rq1/analyze_ablation_jsonl.py --path outputs/rq1/ablation/gsm8k_coconut_gpt2.jsonl
"""

from __future__ import annotations

import argparse
import json
import re
from pathlib import Path
from typing import Any, Dict, List, Optional, Set, Tuple

import matplotlib.pyplot as plt
import numpy as np
from matplotlib.colors import TwoSlopeNorm

from common.experiment_utils import step_order_key

MODE_DESCRIPTIONS: Dict[str, str] = {
    "zero": "Replace latent with zeros",
    "mean": "Replace latent with global mean",
    "gaussian_h": "Add Gaussian noise to latent h_t",
    "gaussian_mu": "Add Gaussian noise around global mean",
    "mean_step": "Replace latent with step-specific mean",
    "gaussian_mu_step": "Add Gaussian noise around step-specific mean",
}


# --------------------------
# Parsing helpers
# --------------------------

def extract_number(text: Any) -> Optional[float]:
    """Extract the LAST number in the answer tail (optionally after <|end-latent|>)."""
    if text is None:
        return None
    s = str(text)
    if "<|end-latent|>" in s:
        s = s.split("<|end-latent|>")[-1]
    s = s.replace(",", "")
    nums = re.findall(r"-?\d+\.?\d*", s)
    if not nums:
        return None
    try:
        return float(nums[-1])
    except Exception:
        return None


def parse_gold(rec: Dict[str, Any]) -> Optional[float]:
    raw = rec.get("gold_answer", "")
    return extract_number(raw)


def normalize_mode_step(mode: Any, step: Any) -> Optional[Tuple[str, int]]:
    """
    Return (mode:str, step:int) with stable labels for sorting/plotting.
    Supports numeric strings/ints -> int.
    """
    if mode is None or step is None:
        return None
    try:
        m = str(mode)

        if isinstance(step, str):
            s = step.strip().lower()
            if s.lstrip("-").isdigit():
                return m, int(s)

        return m, int(step)
    except Exception:
        return None


# --------------------------
# Stats helpers
# --------------------------

def update_acc_stats(
    stats: Dict[Tuple[str, int], Dict[str, float]],
    key: Tuple[str, int],
    gold: Optional[float],
    pred: Optional[float],
) -> None:
    bucket = stats.setdefault(key, {"total": 0.0, "correct": 0.0})
    bucket["total"] += 1.0
    if pred is not None and gold is not None and abs(pred - gold) < 1e-6:
        bucket["correct"] += 1.0


def format_bucket(bucket: Dict[str, float]) -> str:
    total = bucket.get("total", 0.0)
    correct = bucket.get("correct", 0.0)
    acc = correct / total if total else 0.0
    return f"acc={acc:.3f} (correct={int(correct)}/{int(total)})"


def flip_bucket_default() -> Dict[str, float]:
    return {"total": 0.0, "correct_to_wrong": 0.0, "wrong_to_correct": 0.0}


def flip_rates(bucket: Dict[str, float]) -> Tuple[float, float, float, float, float, float]:
    total = bucket.get("total", 0.0)
    w2r = bucket.get("wrong_to_correct", 0.0)
    r2w = bucket.get("correct_to_wrong", 0.0)
    rate = (w2r + r2w) / total if total else 0.0
    w2r_rate = w2r / total if total else 0.0
    r2w_rate = r2w / total if total else 0.0
    return rate, w2r_rate, r2w_rate, total, w2r, r2w


def format_flip(bucket: Dict[str, float]) -> str:
    rate, w2r_rate, r2w_rate, total, w2r, r2w = flip_rates(bucket)
    if total == 0:
        return "flip rate=n/a"
    return (
        f"flip rate={rate:.3f} "
        f"(wrong->right={w2r_rate:.3f} ({int(w2r)}/{int(total)}) "
        f"right->wrong={r2w_rate:.3f} ({int(r2w)}/{int(total)}))"
    )


# --------------------------
# Heatmap summarizers
# --------------------------

def summarize_delta_heatmap(
    ablated_stats: Dict[Tuple[str, int], Dict[str, float]],
    baseline_acc: float,
) -> Tuple[List[str], List[int], List[List[float]], List[List[str]]]:
    modes = sorted({m for (m, _) in ablated_stats.keys()})
    steps = sorted({s for (_, s) in ablated_stats.keys()}, key=step_order_key)

    grid: List[List[float]] = []
    ann: List[List[str]] = []
    for m in modes:
        row: List[float] = []
        ann_row: List[str] = []
        for s in steps:
            b = ablated_stats.get((m, s))
            if not b or b.get("total", 0.0) == 0.0:
                row.append(float("nan"))
                ann_row.append("")
                continue
            acc = b["correct"] / b["total"]
            delta = acc - baseline_acc
            row.append(delta)
            ann_row.append(f"{delta:+.3f}")
        grid.append(row)
        ann.append(ann_row)
    return modes, steps, grid, ann


def summarize_flip_heatmap(
    flip_stats: Dict[Tuple[str, int], Dict[str, float]],
) -> Tuple[List[str], List[int], List[List[float]], List[List[str]]]:
    modes = sorted({m for (m, _) in flip_stats.keys()})
    steps = sorted({s for (_, s) in flip_stats.keys()}, key=step_order_key)

    grid: List[List[float]] = []
    ann: List[List[str]] = []
    for m in modes:
        row: List[float] = []
        ann_row: List[str] = []
        for s in steps:
            b = flip_stats.get((m, s), {})
            rate, _, _, total, _, _ = flip_rates(b)
            if total == 0:
                row.append(float("nan"))
                ann_row.append("")
            else:
                row.append(rate)
                ann_row.append(f"{rate:.2f}")
        grid.append(row)
        ann.append(ann_row)
    return modes, steps, grid, ann


def summarize_logp_heatmap(
    logp_stats: Dict[Tuple[str, int], Dict[str, float]],
    kind: str,
) -> Tuple[List[str], List[int], List[List[float]], List[List[str]]]:
    assert kind in {"seq", "final"}
    modes = sorted({m for (m, _) in logp_stats.keys()})
    steps = sorted({s for (_, s) in logp_stats.keys()}, key=step_order_key)

    grid: List[List[float]] = []
    ann: List[List[str]] = []
    for m in modes:
        row: List[float] = []
        ann_row: List[str] = []
        for s in steps:
            b = logp_stats.get((m, s))
            if not b:
                row.append(float("nan"))
                ann_row.append("")
                continue

            if kind == "seq":
                c = b.get("count_seq", 0.0)
                if c == 0:
                    row.append(float("nan"))
                    ann_row.append("")
                    continue
                val = b["sum_delta_seq"] / c
            else:
                c = b.get("count_final", 0.0)
                if c == 0:
                    row.append(float("nan"))
                    ann_row.append("")
                    continue
                val = b["sum_delta_final"] / c

            row.append(val)
            ann_row.append(f"{val:+.2f}")
        grid.append(row)
        ann.append(ann_row)
    return modes, steps, grid, ann


# --------------------------
# Plot helper
# --------------------------

def plot_heatmap(
    data: List[List[float]],
    modes: List[str],
    steps: List[int],
    title: str,
    cmap: str,
    annotations: List[List[str]],
    out_path: str,
    center: Optional[float] = None,
) -> None:
    arr = np.array(data, dtype=float)
    mask = ~np.isfinite(arr)
    arr_masked = np.ma.array(arr, mask=mask)

    norm = None
    if center is not None:
        finite_vals = arr[np.isfinite(arr)]
        if finite_vals.size:
            max_abs = float(np.max(np.abs(finite_vals)))
            if max_abs > 0:
                norm = TwoSlopeNorm(vcenter=center, vmin=-max_abs, vmax=max_abs)

    figsize = (max(8.0, len(steps) * 1.2), max(4.0, len(modes) * 0.8))
    fig, ax = plt.subplots(figsize=figsize)
    im = ax.imshow(arr_masked, cmap=cmap, norm=norm)

    ax.set_xticks(range(len(steps)))
    ax.set_xticklabels([str(s) for s in steps])
    ax.set_yticks(range(len(modes)))
    ax.set_yticklabels(modes)
    ax.set_xlabel("step")
    ax.set_ylabel("mode")
    ax.set_title(title)

    for i, row in enumerate(annotations):
        for j, text in enumerate(row):
            if not mask[i, j] and text:
                ax.text(j, i, text, ha="center", va="center", fontsize=8)

    fig.colorbar(im, ax=ax, fraction=0.046, pad=0.04)
    fig.tight_layout()
    fig.savefig(out_path, dpi=200)
    plt.close(fig)
    print(f"Saved heatmap to {out_path}")


# --------------------------
# Main
# --------------------------

def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--path", required=True, help="Path to ablation JSONL.")
    args = parser.parse_args()

    baseline_stats: Dict[str, float] = {"total": 0.0, "correct": 0.0}
    baseline_seen: Set[Any] = set()
    baseline_outcomes: Dict[Any, Optional[bool]] = {}  # sid -> correct? (None if unparseable)

    ablated_stats: Dict[Tuple[str, int], Dict[str, float]] = {}
    flip_stats: Dict[Tuple[str, int], Dict[str, float]] = {}
    logp_stats: Dict[Tuple[str, int], Dict[str, float]] = {}

    with open(args.path, "r", encoding="utf-8", errors="replace") as f:
        for line in f:
            if not line.strip():
                continue
            rec = json.loads(line)

            sid = rec.get("sample_id")
            gold_val = parse_gold(rec)

            batch_idx = rec.get("batch_idx", 0)

            # ---- baseline (once per sample_id) ----
            baseline_texts = (rec.get("baseline") or {}).get("text") or []
            if baseline_texts and sid not in baseline_seen:
                text_idx = batch_idx if isinstance(batch_idx, int) else 0
                if text_idx < 0 or text_idx >= len(baseline_texts):
                    text_idx = 0
                pred_val = extract_number(baseline_texts[text_idx])

                baseline_stats["total"] += 1.0
                correct: Optional[bool] = None
                if pred_val is not None and gold_val is not None:
                    correct = abs(pred_val - gold_val) < 1e-6
                    if correct:
                        baseline_stats["correct"] += 1.0
                baseline_outcomes[sid] = correct
                baseline_seen.add(sid)

            # ---- ablated stats per (mode, step) ----
            ms = normalize_mode_step(rec.get("mode"), rec.get("step"))
            if ms is None:
                continue
            mode, step = ms
            key = (mode, step)

            ablated_texts = (rec.get("ablated") or {}).get("text") or []
            if ablated_texts:
                text_idx = batch_idx if isinstance(batch_idx, int) else 0
                if text_idx < 0 or text_idx >= len(ablated_texts):
                    text_idx = 0
                pred_val = extract_number(ablated_texts[text_idx])
                update_acc_stats(ablated_stats, key, gold_val, pred_val)

                # flips
                base_correct = baseline_outcomes.get(sid)
                ablt_correct: Optional[bool] = None
                if pred_val is not None and gold_val is not None:
                    ablt_correct = abs(pred_val - gold_val) < 1e-6

                if base_correct is not None and ablt_correct is not None:
                    fb = flip_stats.setdefault(key, flip_bucket_default())
                    fb["total"] += 1.0
                    if base_correct and not ablt_correct:
                        fb["correct_to_wrong"] += 1.0
                    elif (not base_correct) and ablt_correct:
                        fb["wrong_to_correct"] += 1.0

            # ---- teacher-forced logp deltas ----
            delta_seq = rec.get("teacher_forced_delta_sum", None)
            delta_final = rec.get("delta_logp_final_token", None)

            if delta_seq is not None:
                lb = logp_stats.setdefault(
                    key,
                    {"count_seq": 0.0, "sum_delta_seq": 0.0, "count_final": 0.0, "sum_delta_final": 0.0},
                )
                lb["count_seq"] += 1.0
                lb["sum_delta_seq"] += float(delta_seq)

            if delta_final is not None:
                lb = logp_stats.setdefault(
                    key,
                    {"count_seq": 0.0, "sum_delta_seq": 0.0, "count_final": 0.0, "sum_delta_final": 0.0},
                )
                lb["count_final"] += 1.0
                lb["sum_delta_final"] += float(delta_final)

    # ---- print summaries ----
    print("=== Baseline ===")
    print(format_bucket(baseline_stats))

    baseline_acc = baseline_stats["correct"] / baseline_stats["total"] if baseline_stats["total"] else float("nan")

    print("\n=== Ablated (greedy accuracy & flips) ===")
    modes_present = sorted({m for (m, _) in ablated_stats.keys()})
    if modes_present:
        print("Mode descriptions:")
        for m in modes_present:
            print(f"  {m}: {MODE_DESCRIPTIONS.get(m, 'N/A')}")

    for (mode, step) in sorted(ablated_stats.keys(), key=lambda t: (t[0], step_order_key(t[1]))):
        print(
            f"{mode:15s} step={step}: "
            f"{format_bucket(ablated_stats[(mode, step)])} | "
            f"{format_flip(flip_stats.get((mode, step), flip_bucket_default()))}"
        )

    if logp_stats:
        print("\n=== Teacher-forced logp deltas (base - ablated) ===")
        print("Positive: ablation hurts gold (lower p). Negative: ablation helps gold (higher p).")
        for (mode, step) in sorted(logp_stats.keys(), key=lambda t: (t[0], step_order_key(t[1]))):
            b = logp_stats[(mode, step)]
            seq_msg = "n/a"
            fin_msg = "n/a"
            if b.get("count_seq", 0.0) > 0:
                seq_msg = f"{(b['sum_delta_seq']/b['count_seq']):+.3f} (n={int(b['count_seq'])})"
            if b.get("count_final", 0.0) > 0:
                fin_msg = f"{(b['sum_delta_final']/b['count_final']):+.3f} (n={int(b['count_final'])})"
            print(f"{mode:15s} step={step}: mean Δlogp_seq={seq_msg}, mean Δlogp_final={fin_msg}")

    # ---- heatmaps ----
    prefix = str(Path(args.path).with_suffix(""))

    if ablated_stats:
        rows, steps, grid, ann = summarize_delta_heatmap(ablated_stats, baseline_acc)
        plot_heatmap(
            grid, rows, steps,
            title="Delta accuracy vs baseline",
            cmap="RdBu_r",
            annotations=ann,
            out_path=f"{prefix}_delta_acc_heatmap.png",
            center=0.0,
        )

    if flip_stats:
        rows, steps, grid, ann = summarize_flip_heatmap(flip_stats)
        plot_heatmap(
            grid, rows, steps,
            title="Flip rate (w->r + r->w) by mode/step",
            cmap="plasma",
            annotations=ann,
            out_path=f"{prefix}_flip_heatmap.png",
            center=None,
        )

    if logp_stats:
        rows, steps, grid, ann = summarize_logp_heatmap(logp_stats, kind="seq")
        plot_heatmap(
            grid, rows, steps,
            title="Mean Δ log p(gold sequence) (base - ablt)",
            cmap="RdBu_r",
            annotations=ann,
            out_path=f"{prefix}_delta_logp_seq_heatmap.png",
            center=0.0,
        )

        rows, steps, grid, ann = summarize_logp_heatmap(logp_stats, kind="final")
        plot_heatmap(
            grid, rows, steps,
            title="Mean Δ log p(final gold token) (base - ablt)",
            cmap="RdBu_r",
            annotations=ann,
            out_path=f"{prefix}_delta_logp_final_heatmap.png",
            center=0.0,
        )


if __name__ == "__main__":
    main()

File: r-script/analyze_ablation_jsonl.R
#!/usr/bin/env Rscript

# R port of scripts/plot/python/rq1/analyze_ablation_jsonl.py
# Produces the same heatmaps with the ggplot style used in plot_latent_causal_graph.R.

suppressPackageStartupMessages({
  library(dplyr)
  library(tidyr)
  library(ggplot2)
  library(scales)
})

script_arg <- commandArgs(trailingOnly = FALSE)
script_path <- sub("^--file=", "", script_arg[grep("^--file=", script_arg)])
script_dir <- if (length(script_path) > 0 && nzchar(script_path)) {
  dirname(normalizePath(script_path))
} else {
  normalizePath(".")
}
source(file.path(script_dir, "ablation_parse_utils.R"))

MODE_DESCRIPTIONS <- c(
  "zero" = "Replace latent with zeros",
  "mean" = "Replace latent with global mean",
  "gaussian_h" = "Add Gaussian noise to latent h_t",
  "gaussian_mu" = "Add Gaussian noise around global mean",
  "mean_step" = "Replace latent with step-specific mean",
  "gaussian_mu_step" = "Add Gaussian noise around step-specific mean"
)

to_step_int <- function(x) suppressWarnings(as.integer(as.character(x)))

plot_heatmap <- function(mat, title, fill_label, out_png, diverging = FALSE) {
  if (!nrow(mat)) return(invisible(NULL))
  vals <- mat$value
  if (diverging && any(is.finite(vals))) {
    max_abs <- max(abs(vals[is.finite(vals)]))
    limits <- c(-max_abs, max_abs)
    fill_scale <- scale_fill_distiller(palette = "RdBu", direction = -1, limits = limits, na.value = "grey95")
  } else {
    fill_scale <- scale_fill_viridis_c(option = "C", na.value = "grey95")
  }

  p <- ggplot(mat, aes(x = step, y = mode, fill = value)) +
    geom_tile(color = "white", linewidth = 0.3) +
    geom_text(aes(label = label), size = 3) +
    fill_scale +
    coord_equal() +
    labs(x = "step", y = "mode", fill = fill_label, title = title) +
    theme_minimal(base_size = 12) +
    theme(axis.text.x = element_text(angle = 45, hjust = 1),
          panel.grid = element_blank())

  ggsave(out_png, p, width = 7, height = 6, dpi = 300)
}

parse_args <- function(args) {
  opts <- list(path = NULL, out_dir = NULL)
  i <- 1
  while (i <= length(args)) {
    key <- args[[i]]
    if (key == "--path") {
      opts$path <- args[[i + 1]]
      i <- i + 2
    } else if (key == "--out_dir") {
      opts$out_dir <- args[[i + 1]]
      i <- i + 2
    } else {
      stop(paste("Unknown argument:", key))
    }
  }
  if (is.null(opts$path)) {
    stop("Missing required --path")
  }
  opts
}

args <- commandArgs(trailingOnly = TRUE)
opts <- parse_args(args)
df <- read_jsonl_df(opts$path)

if (!all(c("sample_id", "mode", "step") %in% names(df))) {
  stop("Missing required columns: sample_id, mode, step")
}

df <- df %>%
  mutate(
    step_i = to_step_int(step),
    batch_idx = ifelse("batch_idx" %in% names(df), as.integer(batch_idx), 0L),
    gold_is_choice = grepl("^[A-E]$", as.character(gold_answer)),
    gold = sapply(gold_answer, parse_gold),
    baseline_text = {
      b_list <- to_list_col(baseline, nrow(df))
      vapply(seq_len(nrow(df)), function(i) get_text_field(b_list[[i]], batch_idx[[i]]), character(1))
    },
    ablated_text = {
      a_list <- to_list_col(ablated, nrow(df))
      vapply(seq_len(nrow(df)), function(i) get_text_field(a_list[[i]], batch_idx[[i]]), character(1))
    },
    gold_choice = ifelse(gold_is_choice, as.character(gold_answer), NA_character_),
    gold_num = ifelse(!gold_is_choice, suppressWarnings(as.numeric(gold)), NA_real_),
    baseline_choice = sapply(baseline_text, extract_choice),
    ablated_choice = sapply(ablated_text, extract_choice),
    baseline_pred = sapply(baseline_text, extract_number),
    ablated_pred = sapply(ablated_text, extract_number),
    baseline_correct = ifelse(
      gold_is_choice,
      !is.na(baseline_choice) & !is.na(gold_choice) & baseline_choice == gold_choice,
      !is.na(baseline_pred) & !is.na(gold_num) & abs(baseline_pred - gold_num) < 1e-6
    ),
    ablated_correct = ifelse(
      gold_is_choice,
      !is.na(ablated_choice) & !is.na(gold_choice) & ablated_choice == gold_choice,
      !is.na(ablated_pred) & !is.na(gold_num) & abs(ablated_pred - gold_num) < 1e-6
    )
  )

baseline_tbl <- df %>%
  filter(!is.na(step_i)) %>%
  group_by(step_i) %>%
  summarise(
    total = sum(!is.na(baseline_correct)),
    correct = sum(baseline_correct, na.rm = TRUE),
    acc = ifelse(total > 0, correct / total, NA_real_),
    .groups = "drop"
  )

format_bucket <- function(total, correct) {
  acc <- ifelse(total > 0, correct / total, 0)
  sprintf("acc=%.3f (correct=%d/%d)", acc, as.integer(correct), as.integer(total))
}

format_flip <- function(total, w2r, r2w) {
  if (total <= 0) return("flip rate=n/a")
  rate <- (w2r + r2w) / total
  w2r_rate <- w2r / total
  r2w_rate <- r2w / total
  sprintf(
    "flip rate=%.3f (wrong->right=%.3f (%d/%d) right->wrong=%.3f (%d/%d))",
    rate, w2r_rate, as.integer(w2r), as.integer(total),
    r2w_rate, as.integer(r2w), as.integer(total)
  )
}

ablated_stats <- df %>%
  filter(!is.na(step_i)) %>%
  group_by(mode, step_i) %>%
  summarise(
    total = sum(!is.na(ablated_correct)),
    correct = sum(ablated_correct, na.rm = TRUE),
    acc = ifelse(total > 0, correct / total, NA_real_),
    .groups = "drop"
  ) %>%
  left_join(baseline_tbl %>% select(step_i, baseline_acc = acc), by = "step_i") %>%
  mutate(delta_acc = acc - baseline_acc)

flip_stats <- df %>%
  filter(!is.na(step_i), !is.na(baseline_correct), !is.na(ablated_correct)) %>%
  group_by(mode, step_i) %>%
  summarise(
    total = n(),
    correct_to_wrong = sum(baseline_correct & !ablated_correct, na.rm = TRUE),
    wrong_to_correct = sum(!baseline_correct & ablated_correct, na.rm = TRUE),
    flip_rate = ifelse(total > 0, (correct_to_wrong + wrong_to_correct) / total, NA_real_),
    .groups = "drop"
  )

logp_stats <- df %>%
  filter(!is.na(step_i)) %>%
  group_by(mode, step_i) %>%
  summarise(
    mean_delta_seq = mean(teacher_forced_delta_sum, na.rm = TRUE),
    mean_delta_final = mean(delta_logp_final_token, na.rm = TRUE),
    .groups = "drop"
  )

prefix <- tools::file_path_sans_ext(basename(opts$path))
out_dir <- if (is.null(opts$out_dir)) dirname(opts$path) else opts$out_dir
if (!dir.exists(out_dir)) {
  dir.create(out_dir, recursive = TRUE)
}
prefix <- file.path(out_dir, prefix)

cat("=== Baseline (per step) ===\n")
if (nrow(baseline_tbl) == 0) {
  cat("acc=n/a\n")
} else {
  for (i in seq_len(nrow(baseline_tbl))) {
    row <- baseline_tbl[i, ]
    cat(sprintf("step=%d: %s\n", as.integer(row$step_i), format_bucket(row$total, row$correct)))
  }
}

cat("\n=== Ablated (greedy accuracy & flips) ===\n")
if (nrow(ablated_stats) > 0) {
  modes_present <- sort(unique(ablated_stats$mode))
  cat("Mode descriptions:\n")
  for (m in modes_present) {
    desc <- MODE_DESCRIPTIONS[[m]]
    if (is.null(desc)) desc <- "N/A"
    cat(sprintf("  %s: %s\n", m, desc))
  }

  merged_stats <- ablated_stats %>%
    left_join(flip_stats, by = c("mode", "step_i")) %>%
    arrange(mode, step_i)
  for (i in seq_len(nrow(merged_stats))) {
    row <- merged_stats[i, ]
    acc_msg <- format_bucket(row$total, row$correct)
    flip_msg <- format_flip(row$total.y, row$wrong_to_correct, row$correct_to_wrong)
    cat(sprintf(
      "%-15s step=%d: %s | %s\n",
      as.character(row$mode), as.integer(row$step_i), acc_msg, flip_msg
    ))
  }
}

if (nrow(logp_stats) > 0) {
  cat("\n=== Teacher-forced logp deltas (base - ablt) ===\n")
  cat("Positive: ablation hurts gold (lower p). Negative: ablation helps gold (higher p).\n")
  logp_print <- logp_stats %>%
    arrange(mode, step_i)
  for (i in seq_len(nrow(logp_print))) {
    row <- logp_print[i, ]
    seq_msg <- sprintf("%+.3f (n=%d)", row$mean_delta_seq, nrow(df %>% filter(mode == row$mode, step_i == row$step_i, !is.na(teacher_forced_delta_sum))))
    fin_msg <- sprintf("%+.3f (n=%d)", row$mean_delta_final, nrow(df %>% filter(mode == row$mode, step_i == row$step_i, !is.na(delta_logp_final_token))))
    cat(sprintf(
      "%-15s step=%d: mean Δlogp_seq=%s, mean Δlogp_final=%s\n",
      as.character(row$mode), as.integer(row$step_i), seq_msg, fin_msg
    ))
  }
}

if (nrow(ablated_stats) > 0) {
  mat <- ablated_stats %>%
    mutate(
      mode = factor(mode),
      step = factor(step_i, levels = sort(unique(step_i))),
      label = ifelse(is.na(delta_acc), "", sprintf("%+.3f", delta_acc)),
      value = delta_acc
    ) %>%
    select(mode, step, value, label)
  plot_heatmap(
    mat,
    "Delta accuracy vs baseline",
    "Δacc",
    paste0(prefix, "_delta_acc_heatmap.png"),
    diverging = TRUE
  )
}

if (nrow(flip_stats) > 0) {
  mat <- flip_stats %>%
    mutate(
      mode = factor(mode),
      step = factor(step_i, levels = sort(unique(step_i))),
      label = ifelse(is.na(flip_rate), "", sprintf("%.2f", flip_rate)),
      value = flip_rate
    ) %>%
    select(mode, step, value, label)
  plot_heatmap(
    mat,
    "Flip rate (w->r + r->w) by mode/step",
    "flip_rate",
    paste0(prefix, "_flip_heatmap.png"),
    diverging = FALSE
  )
}

if (nrow(logp_stats) > 0) {
  mat_seq <- logp_stats %>%
    mutate(
      mode = factor(mode),
      step = factor(step_i, levels = sort(unique(step_i))),
      label = ifelse(is.na(mean_delta_seq), "", sprintf("%+.2f", mean_delta_seq)),
      value = mean_delta_seq
    ) %>%
    select(mode, step, value, label)
  plot_heatmap(
    mat_seq,
    "Mean Δ log p(gold sequence) (base - ablt)",
    "Δlogp_seq",
    paste0(prefix, "_delta_logp_seq_heatmap.png"),
    diverging = TRUE
  )

  mat_final <- logp_stats %>%
    mutate(
      mode = factor(mode),
      step = factor(step_i, levels = sort(unique(step_i))),
      label = ifelse(is.na(mean_delta_final), "", sprintf("%+.2f", mean_delta_final)),
      value = mean_delta_final
    ) %>%
    select(mode, step, value, label)
  plot_heatmap(
    mat_final,
    "Mean Δ log p(final gold token) (base - ablt)",
    "Δlogp_final",
    paste0(prefix, "_delta_logp_final_heatmap.png"),
    diverging = TRUE
  )
}
